### PR TITLE
feat: [PLG-1471]: Generalize subscription creation endpoint

### PR DESCRIPTION
**Pull request by thomas-grasman**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/39525)

### CODE REVIEW

**Review by Xin-Harness — DISMISSED** [review]
> Reviewed 10 of 11 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39525)* status: 11 file(s) to review. 1 of 2 approvals obtained.

**Review by kapil-harness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39525)* status: 11 file(s) to review. 1 change requests. 1 of 2 approvals obtained. 

___
*[`930-ng-license-manager/src/main/java/io/harness/subscription/entities/SubscriptionDetail.java` line 44 at r2](https://reviewable.io/reviews/harness/harness-core/39525#-NNMEtEQAPGfn7LOAu3o:-NNMEtERBnnyH1FGaf48:bcdvslh) ([raw file](https://github.com/harness/harness-core/blob/845c7fda9dd224ddecbe604866d1df1e654c994e/930-ng-license-manager/src/main/java/io/harness/subscription/entities/SubscriptionDetail.java#L44)):*
> ```Java
> @Persistent
> public class SubscriptionDetail implements PersistentEntity, NGAccountAccess {
>   public static final String incomplete = "incomplete";
> ```

this variable should be all caps e.g. INCOMPLETE
___
*[`930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/SubscriptionServiceImpl.java` line 226 at r2](https://reviewable.io/reviews/harness/harness-core/39525#-NNMFQzFB3kXop55sBdm:-NNMFQzFB3kXop55sBdn:b-uih2po) ([raw file](https://github.com/harness/harness-core/blob/845c7fda9dd224ddecbe604866d1df1e654c994e/930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/SubscriptionServiceImpl.java#L226)):*
> ```Java
>     isSelfServiceEnable();
> 
>     // TODO: transaction control in case any race condition
> ```

remove these comments. Code should be written in a self explanatory way.
___
*[`935-subscription-sdk/src/main/java/io/harness/subscription/helpers/impl/StripeHelperImpl.java` line 213 at r2](https://reviewable.io/reviews/harness/harness-core/39525#-NNMFoJS56aN2emE3T6I:-NNMFoJS56aN2emE3T6J:b-5cd4jw) ([raw file](https://github.com/harness/harness-core/blob/845c7fda9dd224ddecbe604866d1df1e654c994e/935-subscription-sdk/src/main/java/io/harness/subscription/helpers/impl/StripeHelperImpl.java#L213)):*
> ```Java
>       return priceResult.get();
>     } else {
>       return null;
> ```

do not return NULL rather make priceResult as optional and let the caller of this method handle this
___
*[`950-subscription-beans/src/main/java/io/harness/subscription/params/SubscriptionCreateParams.java` line 20 at r2](https://reviewable.io/reviews/harness/harness-core/39525#-NNMGCYq6Jwl5pTG4_bX:-NNMGCYrAfe5mfAVTzWK:b-mo1gz6) ([raw file](https://github.com/harness/harness-core/blob/845c7fda9dd224ddecbe604866d1df1e654c994e/950-subscription-beans/src/main/java/io/harness/subscription/params/SubscriptionCreateParams.java#L20)):*
> ```Java
> @Builder
> public class SubscriptionCreateParams {
>   private String accountId;
> ```

rename to accountIdentifier
___
*[`950-subscription-beans/src/main/java/io/harness/subscription/params/SubscriptionCreateParams.java` line 23 at r2](https://reviewable.io/reviews/harness/harness-core/39525#-NNMGK-iENAb0RW3W3JM:-NNMGK-iENAb0RW3W3JN:b-wtfuhf) ([raw file](https://github.com/harness/harness-core/blob/845c7fda9dd224ddecbe604866d1df1e654c994e/950-subscription-beans/src/main/java/io/harness/subscription/params/SubscriptionCreateParams.java#L23)):*
> ```Java
>   private ModuleType moduleType;
>   private List<SubscriptionItemParams> items;
>   private String paymentFreq;
> ```

nit: you can write fully qualified name e.g. paymenFrequency

**Review by thomas-grasman — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39525)* status: 13 file(s) to review. 1 change requests. 4. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`930-ng-license-manager/src/main/java/io/harness/subscription/entities/SubscriptionDetail.java` line 44 at r2](https://reviewable.io/reviews/harness/harness-core/39525#-NNMEtEQAPGfn7LOAu3o:-NNOLbBn8RfmLYQfcNhA:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/845c7fda9dd224ddecbe604866d1df1e654c994e/930-ng-license-manager/src/main/java/io/harness/subscription/entities/SubscriptionDetail.java#L44)):*
<details><summary><i>Previously, kapil-harness (Kapil Choudhary) wrote…</i></summary><blockquote>

this variable should be all caps e.g. INCOMPLETE
</blockquote></details>

Done.
___
*[`930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/SubscriptionServiceImpl.java` line 226 at r2](https://reviewable.io/reviews/harness/harness-core/39525#-NNMFQzFB3kXop55sBdm:-NNOLgps0eIZREYskw9i:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/845c7fda9dd224ddecbe604866d1df1e654c994e/930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/SubscriptionServiceImpl.java#L226)):*
<details><summary><i>Previously, kapil-harness (Kapil Choudhary) wrote…</i></summary><blockquote>

remove these comments. Code should be written in a self explanatory way.
</blockquote></details>

Done.
___
*[`935-subscription-sdk/src/main/java/io/harness/subscription/helpers/impl/StripeHelperImpl.java` line 213 at r2](https://reviewable.io/reviews/harness/harness-core/39525#-NNMFoJS56aN2emE3T6I:-NNOLkYP6dXtV-XF3ym2:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/845c7fda9dd224ddecbe604866d1df1e654c994e/935-subscription-sdk/src/main/java/io/harness/subscription/helpers/impl/StripeHelperImpl.java#L213)):*
<details><summary><i>Previously, kapil-harness (Kapil Choudhary) wrote…</i></summary><blockquote>

do not return NULL rather make priceResult as optional and let the caller of this method handle this
</blockquote></details>

Done.
___
*[`950-subscription-beans/src/main/java/io/harness/subscription/params/SubscriptionCreateParams.java` line 20 at r2](https://reviewable.io/reviews/harness/harness-core/39525#-NNMGCYq6Jwl5pTG4_bX:-NNOLnDq8O-wbxIbfs9q:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/845c7fda9dd224ddecbe604866d1df1e654c994e/950-subscription-beans/src/main/java/io/harness/subscription/params/SubscriptionCreateParams.java#L20)):*
<details><summary><i>Previously, kapil-harness (Kapil Choudhary) wrote…</i></summary><blockquote>

rename to accountIdentifier
</blockquote></details>

Done.

**Review by kapil-harness — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39525#-:-NN_78x_0ZV807VlOPas:bnfp4nl)* status: 13 file(s) to review. 1 of 2 approvals obtained.

File: 930-ng-license-manager/src/main/java/io/harness/subscription/services/SubscriptionService.java [hunk 1]
```diff
@@ -38,6 +38,7 @@ SubscriptionDetailDTO updateSubscription(
   boolean checkSubscriptionExists(String subscriptionId);
   List<SubscriptionDetailDTO> listSubscriptions(String accountIdentifier, ModuleType moduleType);
 
+  StripeCustomer getOrCreateCustomer(String accountIdentifier, CustomerDTO customer);
```
Comment (neelam-harness) [inline 930-ng-license-manager/src/main/java/io/harness/subscription/services/SubscriptionService.java]: why do we have single method for two operations `get` or `create` ?

File: 930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/SubscriptionServiceImpl.java [hunk 2]
```diff
@@ -168,105 +170,53 @@ public void payInvoice(String invoiceId) {
...
-  public SubscriptionDetailDTO createFfSubscription(String accountIdentifier, FfSubscriptionDTO subscriptionDTO) {
-    // the module for telemetry event is hard-coded as CF. This method createFfSubscription will be deprecated for
-    // a generic createSubscription where the module won't be hardcoded
-    sendTelemetryEvent("Subscription Creation Initiated", null, accountIdentifier, ModuleType.CF.toString());
-
-    isSelfServiceEnable();
+  private StripeItemRequest buildItemParams(
+      SubscriptionItemRequest itemParams, SubscriptionRequest subscriptionRequest) {
+    Optional<Price> price;
+    StripeItemRequest stripeItemRequestOut = null;
```
Comment (neelam-harness) [inline 930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/SubscriptionServiceImpl.java:176]: Can we just name it as `stripeItemRequest` ? what does `Out` signify here ?

File: 930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/SubscriptionServiceImpl.java [hunk 3]
```diff
@@ -168,105 +170,53 @@ public void payInvoice(String invoiceId) {
...
-  @Override
-  public SubscriptionDetailDTO createFfSubscription(String accountIdentifier, FfSubscriptionDTO subscriptionDTO) {
-    // the module for telemetry event is hard-coded as CF. This method createFfSubscription will be deprecated for
-    // a generic createSubscription where the module won't be hardcoded
-    sendTelemetryEvent("Subscription Creation Initiated", null, accountIdentifier, ModuleType.CF.toString());
-
-    isSelfServiceEnable();
+  private StripeItemRequest buildItemParams(
+      SubscriptionItemRequest itemParams, SubscriptionRequest subscriptionRequest) {
+    Optional<Price> price;
```
Comment (neelam-harness) [inline 930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/SubscriptionServiceImpl.java:175]: this should be renamed as `maybePrice` to represent that it's an optional variable.

File: 930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/SubscriptionServiceImpl.java [hunk 4]
```diff
@@ -168,105 +170,53 @@ public void payInvoice(String invoiceId) {
...
-    if (stripeCustomer == null) {
-      createStripeCustomer(accountIdentifier, subscriptionDTO.getCustomer());
-      stripeCustomer = stripeCustomerRepository.findByAccountIdentifier(accountIdentifier);
+      if (!price.isPresent()) {
+        throw new InvalidArgumentsException(
+            String.format("No price found with metadata: {}, type: {}, edition: {}, billed: {}, max: {}",
+                subscriptionRequest.getModuleType(), itemParams.getType(), subscriptionRequest.getEdition(),
+                subscriptionRequest.getPaymentFrequency(), itemParams.getQuantity()));
+      }
+      stripeItemRequestOut = StripeItemRequest.builder().priceId(price.get().getId()).quantity(1L).build();
```
Comment (neelam-harness) [inline 930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/SubscriptionServiceImpl.java:188]: nit: let's not hardcode `quantity(1L)` and make 1L as a constant instead ?

File: 930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/SubscriptionServiceImpl.java [hunk 5]
```diff
@@ -168,105 +170,53 @@ public void payInvoice(String invoiceId) {
...
+          subscriptionRequest.getEdition(), subscriptionRequest.getPaymentFrequency(), itemParams.getQuantity());
 
-    // verify customer exists
-    StripeCustomer stripeCustomer = stripeCustomerRepository.findByAccountIdentifier(accountIdentifier);
-    if (stripeCustomer == null) {
-      createStripeCustomer(accountIdentifier, subscriptionDTO.getCustomer());
-      stripeCustomer = stripeCustomerRepository.findByAccountIdentifier(accountIdentifier);
+      if (!price.isPresent()) {
+        throw new InvalidArgumentsException(
+            String.format("No price found with metadata: {}, type: {}, edition: {}, billed: {}, max: {}",
```
Comment (neelam-harness) [inline 930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/SubscriptionServiceImpl.java:184]: i see that you have already declared a variable for `PRICE_NOT_FOUND_MESSAGE ` can that not be used here ?

File: 930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/SubscriptionServiceImpl.java (hunk 4 above)
Comment (neelam-harness) [inline 930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/SubscriptionServiceImpl.java:188]: Generally it's better to name the builder methods something like this:
```
StripeItemRequest.builder().withPriceId(abc).withQuantity(123).build();
```

File: 930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/SubscriptionServiceImpl.java [hunk 6]
```diff
@@ -168,105 +170,53 @@ public void payInvoice(String invoiceId) {
...
+                subscriptionRequest.getModuleType(), itemParams.getType(), subscriptionRequest.getEdition(),
+                subscriptionRequest.getPaymentFrequency(), itemParams.getQuantity()));
+      }
+      stripeItemRequestOut = StripeItemRequest.builder().priceId(price.get().getId()).quantity(1L).build();
     } else {
-      updateStripeCustomer(accountIdentifier, stripeCustomer.getCustomerId(), subscriptionDTO.getCustomer());
-    }
-
-    checkEdition(accountIdentifier, subscriptionDTO.getEdition());
+      price = stripeHelper.getPrice(subscriptionRequest.getModuleType(), itemParams.getType(),
```
Comment (neelam-harness) [inline 930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/SubscriptionServiceImpl.java:190]: why don't we just pass down the `subscriptionRequest` in this method and use the required parameters from the request object inside the method itself ? For ex: `stripeHelper.getPrice(subscriptionRequest, itemParams)`

File: 930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/SubscriptionServiceImpl.java [hunk 7]
```diff
@@ -168,105 +170,53 @@ public void payInvoice(String invoiceId) {
...
-          ModuleType.CF, "DEVELOPERS_SUPPORT", subscriptionDTO.getEdition(), subscriptionDTO.getPaymentFreq());
-
-      int maxDevelopersSupport = Integer.parseInt(developerSupportPriceId.getMetadata().getOrDefault("max", "0"));
+    return stripeItemRequestOut;
+  }
 
-      if (subscriptionDTO.getNumberOfDevelopers() > maxDevelopersSupport) {
-        throw new InvalidArgumentsException(QUANTITY_GREATER_THAN_MAX);
-      }
+  private StripeSubscriptionRequest buildSubscriptionParams(
```
Comment (neelam-harness) [inline 930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/SubscriptionServiceImpl.java:205]: rename: `buildSubscriptionRequest`

**Review by neelam-harness — CHANGES_REQUESTED** [review]

**Review by thomas-grasman — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39525)* status: 13 file(s) to review. 1 change requests. 7. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`930-ng-license-manager/src/main/java/io/harness/subscription/services/SubscriptionService.java` line 41 at r4](https://reviewable.io/reviews/harness/harness-core/39525#-NFuTzlLvOWnKDMflVsQ-r4-41:-NNcy85NE3BYys70stBl:ba3dfob) ([raw file](https://github.com/harness/harness-core/blob/aecad62204aafdd13ce9a20d6c1ad831ffb0f56f/930-ng-license-manager/src/main/java/io/harness/subscription/services/SubscriptionService.java#L41)):*
<details><summary><i>Previously, neelam-harness wrote…</i></summary><blockquote>

why do we have single method for two operations `get` or `create` ? 
</blockquote></details>

Removed
___
*[`930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/SubscriptionServiceImpl.java` line 175 at r4](https://reviewable.io/reviews/harness/harness-core/39525#-NFuTzlLvOWnKDMflVsR-r4-175:-NNczZ4rDxcnRLXjHY57:b-6we08g) ([raw file](https://github.com/harness/harness-core/blob/aecad62204aafdd13ce9a20d6c1ad831ffb0f56f/930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/SubscriptionServiceImpl.java#L175)):*
<details><summary><i>Previously, neelam-harness wrote…</i></summary><blockquote>

this should be renamed as `maybePrice` to represent that it's an optional variable. 
</blockquote></details>

Price is expected to exist. If it does not, an exception is thrown.
___
*[`930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/SubscriptionServiceImpl.java` line 176 at r4](https://reviewable.io/reviews/harness/harness-core/39525#-NFuTzlLvOWnKDMflVsR-r4-176:-NNczrJKEsi6T5-IvEeb:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/aecad62204aafdd13ce9a20d6c1ad831ffb0f56f/930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/SubscriptionServiceImpl.java#L176)):*
<details><summary><i>Previously, neelam-harness wrote…</i></summary><blockquote>

Can we just name it as `stripeItemRequest` ? what does `Out` signify here ?
</blockquote></details>

Done.
___
*[`930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/SubscriptionServiceImpl.java` line 184 at r4](https://reviewable.io/reviews/harness/harness-core/39525#-NFuTzlLvOWnKDMflVsR-r4-184:-NNdG_A106VOWyR8be-b:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/aecad62204aafdd13ce9a20d6c1ad831ffb0f56f/930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/SubscriptionServiceImpl.java#L184)):*
<details><summary><i>Previously, neelam-harness wrote…</i></summary><blockquote>

i see that you have already declared a variable for `PRICE_NOT_FOUND_MESSAGE ` can that not be used here ?
</blockquote></details>

Done.
___
*[`930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/SubscriptionServiceImpl.java` line 188 at r4](https://reviewable.io/reviews/harness/harness-core/39525#-NFuTzlLvOWnKDMflVsR-r4-188:-NNdGYVkDh8RmfArqU4u:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/aecad62204aafdd13ce9a20d6c1ad831ffb0f56f/930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/SubscriptionServiceImpl.java#L188)):*
<details><summary><i>Previously, neelam-harness wrote…</i></summary><blockquote>

Generally it's better to name the builder methods something like this:
```
StripeItemRequest.builder().withPriceId(abc).withQuantity(123).build();
```
</blockquote></details>

Done.
___
*[`930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/SubscriptionServiceImpl.java` line 190 at r4](https://reviewable.io/reviews/harness/harness-core/39525#-NFuTzlLvOWnKDMflVsR-r4-190:-NNdGXF9DgmtXvOVd3p1:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/aecad62204aafdd13ce9a20d6c1ad831ffb0f56f/930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/SubscriptionServiceImpl.java#L190)):*
<details><summary><i>Previously, neelam-harness wrote…</i></summary><blockquote>

why don't we just pass down the `subscriptionRequest` in this method and use the required parameters from the request object inside the method itself ? For ex: `stripeHelper.getPrice(subscriptionRequest, itemParams)`
</blockquote></details>

Done.
___
*[`930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/SubscriptionServiceImpl.java` line 205 at r4](https://reviewable.io/reviews/harness/harness-core/39525#-NFuTzlLvOWnKDMflVsR-r4-205:-NNd1Z7U4Rs8xb3LH6rZ:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/aecad62204aafdd13ce9a20d6c1ad831ffb0f56f/930-ng-license-manager/src/main/java/io/harness/subscription/services/impl/SubscriptionServiceImpl.java#L205)):*
<details><summary><i>Previously, neelam-harness wrote…</i></summary><blockquote>

rename: `buildSubscriptionRequest`
</blockquote></details>

Done.



_Code quote:_
```Java
  private StripeSubscriptionRequest buildSubscriptionParams(
```

File: 935-subscription-sdk/src/main/java/io/harness/subscription/helpers/impl/StripeHelperImpl.java [hunk 8]
```diff
@@ -198,20 +200,14 @@ public PriceCollectionDTO getPrices(ModuleType moduleType) {
...
-    return stripeHandler.searchPrices(params).getData().stream().findFirst();
-  }
-
-  @Override
-  public Optional<Price> getPrice(ModuleType moduleType, String type, String edition, String paymentFrequency) {
-    String searchString =
-        String.format(SEARCH_MODULE_TYPE_EDITION_BILLED, moduleType.toString(), type, edition, paymentFrequency);
+      SubscriptionRequest subscriptionRequest, SubscriptionItemRequest subscriptionItemRequest) {
+    String searchString = subscriptionItemRequest.isQuantityIncludedInPrice()
+        ? String.format(SEARCH_MODULE_TYPE_EDITION_BILLED_MAX, subscriptionRequest.getModuleType().toString(),
```
Comment (neelam-harness) [inline 935-subscription-sdk/src/main/java/io/harness/subscription/helpers/impl/StripeHelperImpl.java:205]: we can create a helper method here and pass in the variable. Something like this:

```
String searchString = subscriptionItemRequest.isQuantityIncludedInPrice()
     ? buildSearchString(SEARCH_MODULE_TYPE_EDITION_BILLED_MAX, subscriptionRequest) : buildSearchString(SEARCH_MODULE_TYPE_EDITION_BILLED, subscriptionRequest);


private String buildSearchString(String searchModuleType, SubscriptionRequest subscriptionRequest) {
      return String.format(searchModuleType, subscriptionRequest.getModuleType().toString(),
            subscriptionItemRequest.getType(), subscriptionRequest.getEdition(),
            subscriptionRequest.getPaymentFrequency());
}
```

**Review by neelam-harness — CHANGES_REQUESTED** [review]

**Review by thomas-grasman — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39525)* status: 13 file(s) to review. 1 change requests. 8. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`935-subscription-sdk/src/main/java/io/harness/subscription/helpers/impl/StripeHelperImpl.java` line 205 at r6](https://reviewable.io/reviews/harness/harness-core/39525#-NFuTzlLvOWnKDMflVsT-r6-205:-NNeHZau3IfiFJOHFNKz:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/72ee94305e228369822d0fcd4f501329b70ae342/935-subscription-sdk/src/main/java/io/harness/subscription/helpers/impl/StripeHelperImpl.java#L205)):*
<details><summary><i>Previously, neelam-harness wrote…</i></summary><blockquote>

we can create a helper method here and pass in the variable. Something like this:

```
String searchString = subscriptionItemRequest.isQuantityIncludedInPrice()
     ? buildSearchString(SEARCH_MODULE_TYPE_EDITION_BILLED_MAX, subscriptionRequest) : buildSearchString(SEARCH_MODULE_TYPE_EDITION_BILLED, subscriptionRequest);


private String buildSearchString(String searchModuleType, SubscriptionRequest subscriptionRequest) {
      return String.format(searchModuleType, subscriptionRequest.getModuleType().toString(),
            subscriptionItemRequest.getType(), subscriptionRequest.getEdition(),
            subscriptionRequest.getPaymentFrequency());
}
```




</blockquote></details>

Done.

**Review by kapil-harness — APPROVED** [review]
> Reviewed 1 of 11 files at r1, 5 of 11 files at r3, 2 of 5 files at r4, 2 of 5 files at r6, 2 of 2 files at r7, 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39525)* status: 12 file(s) to review. 1 change requests. 8. unresolved discussions. 1 of 2 approvals obtained.

**Review by neelam-harness — APPROVED** [review]
> Reviewed 1 of 11 files at r1, 5 of 11 files at r3, 2 of 5 files at r4, 2 of 5 files at r6, 2 of 2 files at r7, 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39525)* status: :shipit: complete! ✓

File: 950-subscription-beans/src/main/java/io/harness/subscription/params/SubscriptionRequest.java [hunk 9]
```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright 2021 Harness Inc. All rights reserved.
```
Comment (neelam-harness) [inline 950-subscription-beans/src/main/java/io/harness/subscription/params/SubscriptionRequest.java:2]: nit: let's change this from `2021` to `2023` ?

**Pull request merged**